### ISSUE TITLE
usegeocache with incremental node construction

### ISSUE THREAD

**Issue by andreasplesch** [x3dom/x3dom#967]
The default of using the mesh cache by setting usegeocache true is often problematic when geometry nodes are constructed with d3. There, typically, first a default node (say Cylinder) is appended which causes x3dom to cache the default mesh for the geometry. Then, step by step, attributes of the node are modified to the desired values which causes the cached mesh to change.

Since the default node is always used, subsequent addition of the same node type (Cylinder) cause x3dom to reuse the cached mesh by reference, not by cloning. Referencing normally would be beneficial but in this scenario, the mesh caches becomes out of sync with unexpected results:

https://github.com/jamesleesaunders/d3-x3dom/issues/84#issuecomment-505118829

There are several options:

- opt in on usegeocache, eg. make default false

This would let scene designers better control use of the cache since now there is no way to switch it off with d3. But in all other cases, default use would be preferred.

- fix cache coherency in fieldChanged

This may be the proper fix but requires a larger effort in most geometry nodes. Not sure exactly how. One way may to always clone a cached mesh before assigning it to the node. But this would negate some of the benefits of the cache.

**Comment by andreasplesch** [comment]
I think the underlying problem is that usegeocache cannot be set by adding the attribute later. usegeocache is designed to avoid using the cache exactly for these situations and so should be made available.

I think something like this may work:

in fieldchanged:
```js
if (fieldName === "useGeoCache")
            	{
            		if (this._vf.useGeoCache === false)
            		{
            			this._mesh = this._mesh.copy();
            		}
            	}
```
in Mesh.js
```js
x3dom.Mesh.prototype.copy = function() {
	var copy = new x3dom.Mesh(this.parent);
	Object.assign(copy, this);
	copy._positions = Array.from(this._positions);
    copy._normals = Array.from(this._normals);
    copy._texCoords = Array.from(this._texCoords);
    copy._texCoords2 = Array.from(this._texCoords2);
    copy._colors    = Array.from(this._colors);
    copy._indices   = Array.from(this._indices);
    copy._tangents  = Array.from(this._tangents);
    copy._binormals = Array.from(this._binormals);

    copy._positions[0] = Array.from(this._positions[0]);
    copy._normals[0] = Array.from(this._normals[0]);
    copy._texCoords[0] = Array.from(this._texCoords[0]);
    copy._texCoords2[0] = Array.from(this._texCoords2[0]);
    copy._colors[0]    = Array.from(this._colors[0]);
    copy._indices[0]   = Array.from(this._indices[0]);
    copy._tangents[0]  = Array.from(this._tangents[0]);
    copy._binormals[0] = Array.from(this._binormals[0]);
    Object.assign(copy._vol, this._vol);
    return copy
};
```

Then one can use d3selection.append('Cylinder').attr('usegeocache','false')

**Comment by jamesleesaunders** [comment]
Would this flag be more appropriate at a more global level?
`<x3d ... useGeoCashe=‘false’>`
Defaulted to true of course, then it would work for all shapes?
Of course I would assume this would come with some performance degradation?

**Comment by andreasplesch** [comment]
Yeah, I had that thought, too. It would be an option to set the default useGeoCache value, sort of similar to primitive quality.

**Comment by andreasplesch** [comment]
@jamesleesaunders : I have added a global useGeoCache option to a usegeocache branch:

https://github.com/andreasplesch/x3dom/tree/usegeocache

`<X3D useGeoCache='false' ..>` should set the default usegeocache value and so disable the cache globally.

https://5d19fc3dfe3989fb715db854--x3dom.netlify.com/ has a build of this branch.

**Comment by jamesleesaunders** [comment]
@andreasplesch This looks to work well. But I think there may be a typo in X3DOMGeometryNode as at the moment useGeoCache looks to be permanently false regardless of `<X3D useGeoCache='false' ..>` or `<X3D useGeoCache='true' ..>`. 

I think I spotted the typo see https://github.com/andreasplesch/x3dom/pull/141

I have tested this version with d3-x3dom and the axis ticks are now behaving (no artefacts) when useGeoCache is `false` and the ticks are misbehaving (expectedly) when `true`.
